pkg/common/cns-lib/volume/manager.go

```diff
@@ -565,63 +576,184 @@ func (m *defaultManager) DeleteVolume(ctx context.Context, volumeID string, dele
+	taskResult, err := getTaskResultFromTaskInfo(ctx, taskInfo)
+	if err != nil {
+		log.Errorf("failed to get DeleteVolume task result with error: %v", err)
+		return err
+	}
+	volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
+	if volumeOperationRes.Fault != nil {
+		return logger.LogNewErrorf(log, "failed to delete volume: %q, fault: %q, opID: %q",
+			volumeID, spew.Sdump(volumeOperationRes.Fault), taskInfo.ActivationId)
+	}
+	log.Infof("DeleteVolume: Volume deleted successfully. volumeID: %q, opId: %q",
+		volumeID, taskInfo.ActivationId)
+	return nil
+}
+
+// deleteVolumeWithIdempotency attempts to delete the volume on CNS. CNS task information is persisted by leveraging
+// the VolumeOperationRequest interface.
+func (m *defaultManager) deleteVolumeWithIdempotency(ctx context.Context, volumeID string, deleteDisk bool) error {
+	log := logger.GetLogger(ctx)
+	var (
+		// Reference to the DeleteVolume task on CNS.
+		task *object.Task
+		// Name of the CnsVolumeOperationRequest instance.
+		instanceName = "delete-" + volumeID
+		// Local instance of DeleteVolume details that needs to
+		// be persisted.
+		volumeOperationDetails *cnsvolumeoperationrequest.VolumeOperationRequestDetails
+	)
+	if m.operationStore == nil {
+		return logger.LogNewError(log, "operation store cannot be nil")
+	}
+	// Determine if CNS needs to be invoked.
+	volumeOperationDetails, err := m.operationStore.GetRequestDetails(ctx, instanceName)
+	switch {
+	case err == nil:
+		if volumeOperationDetails.OperationDetails != nil {
+			// Validate if previous attempt was
+			// successful.
+			if volumeOperationDetails.OperationDetails.TaskStatus == taskInvocationStatusSuccess {
+				return nil
+			}
+			// Validate if previous operation is
+			// pending.
+			if volumeOperationDetails.OperationDetails.TaskStatus == taskInvocationStatusInProgress &&
+				volumeOperationDetails.OperationDetails.TaskID != "" {
+				taskMoRef := vim25types.ManagedObjectReference{
+					Type:  "Task",
+					Value: volumeOperationDetails.OperationDetails.TaskID,
+				}
+				task = object.NewTask(m.virtualCenter.Client.Client, taskMoRef)
+			}
+		}
+	case apierrors.IsNotFound(err):
+		volumeOperationDetails = createRequestDetails(instanceName, "", "", 0, metav1.Now(),
+			"", "", taskInvocationStatusInProgress, "")
+	default:
+		return err
+	}
+
+	defer func() {
+		// Persist the operation details before returning. Only success or error needs to be stored as
+		// InProgress details are stored when the task is created on CNS.
+		if volumeOperationDetails != nil && volumeOperationDetails.OperationDetails != nil &&
+			volumeOperationDetails.OperationDetails.TaskStatus != taskInvocationStatusInProgress {
+			err := m.operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
+			if err != nil {
+				log.Warnf("failed to store DeleteVolume details with error: %v", err)
+			}
+		}
+	}()
+
+	if task == nil {
+		// No previos DeleteVolume task was persisted for this volume.
```

Comment:
Nit: We also have a case where the previous task may have failed.

Reply:
done